src/kdc/kdc_authdata.c

```diff
@@ -792,8 +376,221 @@ get_auth_indicators(krb5_context context, krb5_enc_tkt_part *enc_tkt,
+        if (ret)
+            goto done;
+
+        ret = k5_ndr_dec_delegation_info(&data, &di);
+        if (ret)
+            goto done;
+
+        /* Update target to proxy->princ. */
+        ret = krb5_unparse_name(context, client->princ, &tmpstr);
+        if (ret)
+            goto done;
+
+        payload = k5_ndr_enc_rpc_unicode_string(tmpstr);
+        free(tmpstr);
+        if (payload == NULL)
+            goto done;
+        free(di->proxy_target);
+        di->proxy_target = payload;
+
+        /* Add entry to transited services. */
+        ret = krb5_unparse_name(context, server->princ, &tmpstr);
+        if (ret)
+            goto done;
+
+        payload = k5_ndr_enc_rpc_unicode_string(tmpstr);
+        free(tmpstr);
+        if (payload == NULL)
+            goto done;
+        di->transited_services[di->transited_services_length++] = payload;
+
+        ret = k5_ndr_enc_delegation_info(di, &data);
+        if (ret)
+            goto done;
+
+        ret = krb5_pac_add_buffer(context, new_pac, KRB5_PAC_DELEGATION_INFO,
+                                  &data);
+        if (ret)
+            goto done;
+    }
+
+done:
+    krb5_free_data_contents(context, &data);
+    k5_ndr_free_delegation_info(&di);
+    /* Caller takes care of freeing new_pac. */
+    return ret;
+}
+
+/*
+ * MS-KILE (3.3.5.3) says to issue a PAC only when requested (but that clients
+ * will always request in AS-REQ), and either:
+ *
+ *   - AS-REQ which preauth and for which PAC generation isn't disabled
+ *   - successful TGS-REQ when it wasn't disabled in the incoming PAC
+ *
+ * MS-KILE (3.3.5.7.2) says that TGS-REQ from a TGT without a PAC "SHOULD"
+ * result in a service ticket with a fresh PAC.  (3.3.5.7) says that
+ * otherwise, the PAC fields are copied from the TGT.  (Signatures are
+ * updated, though it does not mention this.)
+ *
+ * So if the incoming request had one (TGS), we verify it and attach it,
+ * resigned.  Otherwise, we ask the KDB for a fresh one.
+ */
+static krb5_error_code
+handle_pac(krb5_context context, unsigned int flags, unsigned int kdc_options,
+           krb5_db_entry *client, krb5_db_entry *server,
+           krb5_db_entry *subject_server, krb5_db_entry *local_tgt,
+           krb5_keyblock *local_tgt_key, krb5_keyblock *server_key,
+           krb5_const_principal altcprinc, krb5_enc_tkt_part *enc_tkt_reply,
+           krb5_ticket *old_ticket, krb5_ticket *new_ticket)
+{
+    krb5_error_code ret;
+    krb5_pac old_pac = NULL, new_pac = NULL;
+    krb5_boolean is_cross;
+    krb5_const_principal client_princ = NULL;
+    krb5_db_entry *krbtgt = local_tgt;
+    krb5_keyblock *krbtgt_key = local_tgt_key;
+
+    /* Select the right krbtgt for the realm. */
+    if (server_key != NULL)
+        krbtgt_key = server_key;
+    if (subject_server != NULL)
+        krbtgt = subject_server;
+
+    is_cross = is_cross_tgs_principal(krbtgt->princ);
+
+    old_pac = get_pac(context, enc_tkt_reply->authorization_data);
```

Comment:
enc_tkt_reply is the encrypted part of the ticket we're issuing.  I don't think we'd ever find an old PAC there.